pennylane_lightning/src/util/Kokkos_Sparse.hpp

```diff
@@ -0,0 +1,179 @@
+ * @param entries_ptr   Pointer to the array with the non-zero elements column
+ * indices.
+ * @param values_ptr    Pointer to the array with the non-zero elements.
+ * @param numNNZ        Number of non-zero elements.
+ * @return crs_matrix_type
+ */
+template <class fp_precision>
+const_crs_matrix_type<fp_precision> create_Kokkos_Sparse_Matrix(
+    const index_type *row_map_ptr, const index_type numRows,
+    const index_type *entries_ptr, const std::complex<fp_precision> *values_ptr,
+    const index_type numNNZ) {
+    const_index_view_type row_map(row_map_ptr, numRows + 1);
+    const_index_view_type entries(entries_ptr, numNNZ);
+    const_data_view_type<fp_precision> values(
+        reinterpret_cast<const Kokkos::complex<fp_precision> *>(values_ptr),
+        numNNZ);
+
+    const_graph_type<fp_precision> myGraph(entries, row_map);
+    const_crs_matrix_type<fp_precision> SparseMatrix("matrix", numRows, values,
+                                                     myGraph);
+    return SparseMatrix;
+}
+
+/**
+ * @brief Apply a sparse matrix to a vector.
+ *
+ * @tparam fp_precision
+ * @param vector_ptr    pointer to the vector.
+ * @param vector_size   size of the vector.
+ * @param row_map_ptr   pointer to the row_map.
+ *                      The j element encodes the number of non-zeros above row
+ * j.
+ * @param row_map_size  number of elements in the row_map.
+ * @param entries_ptr   pointer to the column indices of the non-zero elements.
+ * @param values_ptr    non-zero elements.
+ * @param numNNZ        number of non-zero elements.
+ * @return std::vector<data_type<fp_precision>>
+ */
+template <class fp_precision>
+std::vector<std::complex<fp_precision>> apply_Sparse_Matrix(
+    const std::complex<fp_precision> *vector_ptr, const index_type vector_size,
+    const index_type *row_map_ptr, const index_type row_map_size,
+    const index_type *entries_ptr, const std::complex<fp_precision> *values_ptr,
+    const index_type numNNZ) {
+
+    std::vector<std::complex<fp_precision>> result;
+    Kokkos::initialize();
+    {
+        const_data_view_type<fp_precision> vector_view(
+            reinterpret_cast<const Kokkos::complex<fp_precision> *>(vector_ptr),
+            vector_size);
+        Kokkos::View<data_type<fp_precision> *> result_view("rhs", vector_size);
+
+        const_crs_matrix_type<fp_precision> myMatrix =
+            create_Kokkos_Sparse_Matrix(row_map_ptr, row_map_size - 1,
+                                        entries_ptr, values_ptr, numNNZ);
+
+        const data_type<fp_precision> alpha(1.0);
+        const data_type<fp_precision> beta;
+        KokkosSparse::spmv("N", alpha, myMatrix, vector_view, beta,
+                           result_view);
+
+        result = std::move(std::vector<std::complex<fp_precision>>(
+            reinterpret_cast<std::complex<fp_precision> *>(result_view.data()),
+            reinterpret_cast<std::complex<fp_precision> *>(result_view.data()) +
```

Comment:
I assume this reinterpret is due to the Kokkos complex vs std::complex issues we discussed previously?